earth_enterprise/src/fusion/autoingest/sysman/StateUpdater.cpp

```diff
@@ -278,16 +291,86 @@ class StateUpdater::SetStateVisitor : public default_dfs_visitor {
     }
 };
 
-void StateUpdater::SetStateForRefAndDependents(
+
+class StateUpdater::SetBlockingStateVisitor : public StateUpdater::VisitorBase {
+  private:
+    using BlockedSet = std::unordered_set<SharedString>;
+
+    // This is a shared_ptr because the visitor will be copied several times
+    // during the course of the traversal.
+    const std::shared_ptr<BlockedSet> hasBlockingInputs;
+    const std::shared_ptr<BlockedSet> hasBlockingChildren;
+
+    void MarkParentsForLaterProcessing(AssetHandle<AssetVersionImpl> & version) const override {
+      hasBlockingChildren->insert(version->parents.begin(), version->parents.end());
+      hasBlockingInputs->insert(version->listeners.begin(), version->listeners.end());
+    }
+
+  public:
+    SetBlockingStateVisitor(StateUpdater & updater, DependentStateTree & tree, AssetDefs::State newState) :
+        VisitorBase(updater, tree, newState),
+        hasBlockingInputs(std::make_shared<BlockedSet>()),
+        hasBlockingChildren(std::make_shared<BlockedSet>()) 
+        { assert(AssetDefs::IsBlocking(newState)); }
+
+    // This function is called after the DFS has completed for every vertex
+    // below this one in the tree. Thus, we don't calculate the state for an
+    // asset until we've calculated the state for all of its inputs and
+    // children.
+    void finish_vertex(
+        DependentStateTreeVertexDescriptor vertex,
+        const DependentStateTree &) const {
+      const AssetVertex & data = tree[vertex];
+      notify(NFY_PROGRESS, "Calculating state for '%s'", data.name.toString().c_str());
+
+      // Set the state for assets in the dependent tree.
+      if (data.inDepTree) {
+        SetState(vertex, newState, {0, 0}, true);
+      }
+      else {
+        // Set the state to blocked if needed, otherwise do nothing
+        auto version = updater.storageManager->Get(data.name);
+        auto inputIt = hasBlockingInputs->find(data.name);
+        if (inputIt != hasBlockingInputs->end() && 
+            version->InputStatesAffectMyState(AssetDefs::Blocked, true)) {
+          SetState(vertex, AssetDefs::Blocked, {0,0}, true);
+          hasBlockingInputs->erase(inputIt);
```

Comment:
Minor: if the asset is in `hasBlockingInputs` but `InputStatesAffectMyState` returns false then this line will never be called, so some stuff may not get removed from the `hasBlockingInputs`. Not a huge deal, but not hard to fix either.

Reply:
Good catch. I completely missed that.